src/DotNetty.Handlers/Tls/SniHandler.cs

```diff
@@ -0,0 +1,260 @@
+                                                //    PlatformDependent.throwException(t);
+                                                //}
+                                                this.Select(context, hostname.ToLowerInvariant()); 
+                                                return;
+                                            }
+                                            else
+                                            {
+                                                // invalid enum value
+                                                continueLoop = false;
+                                                break;
+                                            }
+                                        }
+
+                                        offset += extensionLength;
+                                    }
+                                }
+
+                                break;
+                            // Fall-through
+                            default:
+                                //not tls, ssl or application data, do not try sni
+                                continueLoop = false;
+                                break;
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    error = e;
+
+                    // unexpected encoding, ignore sni and use default
+                    if (Logger.DebugEnabled)
+                    {
+                        Logger.Warn($"Unexpected client hello packet: {ByteBufferUtil.HexDump(input)}", e);
+                    }
+                }
+
+                if (this.serverTlsSniSettings.DefaultServerHostName != null)
+                {
+                    // Just select the default certifcate
+                    this.Select(context, this.serverTlsSniSettings.DefaultServerHostName); 
+                }
+                else
+                {
+                    this.handshakeFailed = true;
+                    var e = new InvalidOperationException("Unable to determine server certicate", error);
+                    TlsUtils.NotifyHandshakeFailure(context, e);
+                    throw e;
+                }
+            }
+        }
+
+        void Select(IChannelHandlerContext context, string hostName) => this.ReplaceHandler(context, hostName);
+
+        void ReplaceHandler(IChannelHandlerContext context, string hostName)
+        {
+            Contract.Requires(hostName != null);
+            var serverTlsSetting = new ServerTlsSettings(this.serverTlsSniSettings.ServerCertificateSelector(hostName), this.serverTlsSniSettings.NegotiateClientCertificate, this.serverTlsSniSettings.CheckCertificateRevocation, this.serverTlsSniSettings.EnabledProtocols);
+            var tlsHandler = new TlsHandler(this.sslStreamFactory, serverTlsSetting);
+            context.Channel.Pipeline.Replace(this, nameof(TlsHandler), tlsHandler);
+        }
+    }
```

Comment:
Okay the reason was suppress read is used in netty during the async sslcontext lookup being executed. In this case so far I was using sync cert lookup. But now since I changed the cert lookup async now as per your other comment, I have now ported the same read of netty.